### PR TITLE
This changes the generator from a conv scope generator to a conv cache iterator

### PR DESCRIPTION
**Pull request by chuckbutler**
This allows the user to consume the list_unit_data method as iterating
the cache of relation data

for example:

    @when('elasticsearch.available')
          def stuff_with_es(elasticsearch):
                for unit in elasticsearch.list_unit_data():
                print(unit['host'])

### CODE REVIEW

File: requires.py [hunk 1]
```diff
@@ -28,19 +28,22 @@ def changed(self):
...
-        self.get_remote('cluster-name')
-
-    def list_units(self):
+        for unit in elasticsearch.list_data():
+            print(unit['cluster_name'])
+        '''
         for conv in self.conversations():
-            unit = conv.scope
-            yield unit
+            yield {'cluster_name': conv.get_remote('cluster_name'),
```
Comment (mbruzek) [inline requires.py]: I don't know where this would come from. It is not an auto accessor.

Comment (chuckbutler) [inline requires.py]: this comes from the remote data.

So as I understand it, and feel free to correct me @johnsca  or @marcoceppi :+1: 

Interface layers cache data that is explicitly sent across the wire with relation-set, the auto_accessors are convenience method(s) for globally scoped relationship interfaces. This is a scope:unit so we need to iterate the conversation objects to retrieve the data, similar to how we used a scope={{conv scope}} while setting relation data on a non-globally scoped relation.

So, in short, the data is coming from elasticsearch, which is relation-setting the data. we dont get handy accessors because the scope:unit of the relationship doesn't retain that self. bit, unless its implicitly in the interface class, during the hook sequence that the conversation is assigned to self.

or something along those lines.

Comment (johnsca) [inline requires.py]: Correct.  `auto_accessors` is only valid for `GLOBAL` scope; it is equivalent to `self.conversation().get_remote(attr)` and depends on a default conversation, which only works if you happen to be in a hook context (not something that you should ever rely on unless you're in a `@hook` handler) or only have exactly 1 conversation (i.e., `GLOBAL` scope).  Otherwise, calling an `auto_accessor` method (or `self.get_remote()`) will fail with an error.

For `SERVICE` or `UNIT` scopes, you must explicitly loop over the conversations which will represent either all units of a service (`SERVICE` scope), or a single unit (`UNIT` scope).  For `UNIT` scope, `conv.get_remote(key)` is equivalent to `relation_get(key)` with the correct value for the `unit` and `rel_id` params.  For `SERVICE` scope, it will pick the value set by any unit of that service which has set a value, so they should either all (eventually) agree or the leader should set the value.
